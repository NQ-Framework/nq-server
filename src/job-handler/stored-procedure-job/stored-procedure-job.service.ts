import { Injectable } from '@nestjs/common';
import {
  ScheduledJob,
  StoredProcedureConfiguration,
} from '@nqframework/models';
import { BaseJobService } from '../base-job/base-job.service';

@Injectable()
export class StoredProcedureJobService extends BaseJobService {
  ExecuteJob(config: StoredProcedureConfiguration, job: ScheduledJob): void {
    console.log(
      'this would execute a job called ' + job.name + 'sp for org ' + job.organizationId + ' with: ',
      JSON.stringify(config),
    );
  }
}
