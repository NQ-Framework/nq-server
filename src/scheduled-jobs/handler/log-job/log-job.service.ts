import { Injectable } from '@nestjs/common';
import { LogConfiguration } from '@nqframework/models';
import { BaseJobService } from '../base-job/base-job.service';

@Injectable()
export class LogJobService extends BaseJobService {
    ExecuteJob(config: LogConfiguration): void {
        console.log(config.message);
    }
}
