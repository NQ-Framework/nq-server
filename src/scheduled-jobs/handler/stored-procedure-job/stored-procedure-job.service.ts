import { Injectable } from '@nestjs/common';
import { StoredProcedureConfiguration } from '@nqframework/models';
import { BaseJobService } from '../base-job/base-job.service';

@Injectable()
export class StoredProcedureJobService extends BaseJobService {
    ExecuteJob(config: StoredProcedureConfiguration): void {
        console.log("this would execute a sp with: ", JSON.stringify(config));
    }
}
