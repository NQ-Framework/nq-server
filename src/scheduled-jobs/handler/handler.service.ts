import { Injectable } from '@nestjs/common';
import { ConfigurationInterface } from '@nqframework/models';
import { BaseJobService } from './base-job/base-job.service';
import { LogJobService } from "./log-job/log-job.service";
import { StoredProcedureJobService } from './stored-procedure-job/stored-procedure-job.service';


@Injectable()
export class HandlerService {
    map: any = {};
    constructor(logJob: LogJobService, storedProcedureJob: StoredProcedureJobService) {
        this.map['log'] = logJob;
        this.map['stored-procedure'] = storedProcedureJob;
    }

    GetHandlerFromConfig(config: ConfigurationInterface): BaseJobService | null {
        const handler = this.map[config.type];
        return handler ? handler : null;
    }
}
