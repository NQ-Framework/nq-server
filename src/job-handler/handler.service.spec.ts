import { Test, TestingModule } from '@nestjs/testing';
import { LoggerModule } from '../logger/logger.module';
import { HandlerService } from './handler.service';
import { LogJobService } from './log-job/log-job.service';
import { StoredProcedureJobService } from './stored-procedure-job/stored-procedure-job.service';

describe('HandlerService', () => {
  let service: HandlerService;
  const executeJobMock = jest.fn().mockImplementation(() => 'return data');

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [LoggerModule],
      providers: [
        HandlerService,
        {
          provide: LogJobService, useValue: {
            ExecuteJob: executeJobMock
          }
        },
        { provide: StoredProcedureJobService, useValue: {} },
      ],
    }).compile();

    service = module.get<HandlerService>(HandlerService);
  });

  it('should fetch handlers', () => {
    expect(service.getHandlerFromConfig({ type: 'log' })).toBeDefined();
    expect(service.getHandlerFromConfig({ type: 'stored-procedure' })).toBeDefined();
    expect(service.getHandlerFromConfig({ type: 'invalid' } as any)).toBeNull();
  });

  it('should execute jobs', async () => {
    const data = await service.executeJob({ configuration: { type: 'log' }, name: 'mock job' } as any);
    expect(executeJobMock).toHaveBeenCalledWith({ type: "log" }, expect.objectContaining({
      name: 'mock job'
    }));
    expect(data).toEqual({ data: 'return data' });
  });

  it('should throw on invalid job configuration', async () => {
    await expect(service.executeJob({ configuration: { type: 'invalid type' }, name: 'mock job' } as any)).rejects.toThrowErrorMatchingSnapshot();
  });
});
