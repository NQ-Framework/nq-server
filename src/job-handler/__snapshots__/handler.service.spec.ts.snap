// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HandlerService should throw on invalid job configuration 1`] = `"handler not found for job"`;
